File: version_3/code-samples/Indexes/SpatialIndexes.cs
using System.Linq;

using Raven.Abstractions.Indexing;
using Raven.Client;
using Raven.Client.Document;
using Raven.Client.Indexes;
using Raven.Client.Spatial;

namespace Raven.Documentation.CodeSamples.Indexes
{
	namespace Foo
	{

		#region spatial_search_enhancements_3
		public class SpatialOptionsFactory
		{
			public GeographySpatialOptionsFactory Geography;

			public CartesianSpatialOptionsFactory Cartesian;
		}

		#endregion

		public interface GeographySpatialOptionsFactory
		{
			#region spatial_search_enhancements_4
			// GeohashPrefixTree strategy with maxTreeLevel set to 9
			SpatialOptions Default(SpatialUnits circleRadiusUnits = SpatialUnits.Kilometers);

			SpatialOptions BoundingBoxIndex(SpatialUnits circleRadiusUnits = SpatialUnits.Kilometers);

			SpatialOptions GeohashPrefixTreeIndex(int maxTreeLevel, SpatialUnits circleRadiusUnits = SpatialUnits.Kilometers);

			SpatialOptions QuadPrefixTreeIndex(int maxTreeLevel, SpatialUnits circleRadiusUnits = SpatialUnits.Kilometers);

			#endregion
		}

		public interface SpatialCriteriaFactory
		{
			#region spatial_search_enhancements_a
			SpatialCriteria RelatesToShape(object shape, SpatialRelation relation);

			SpatialCriteria Intersects(object shape);

			SpatialCriteria Contains(object shape);

			SpatialCriteria Disjoint(object shape);

			SpatialCriteria Within(object shape);

			SpatialCriteria WithinRadiusOf(double radius, double x, double y);

			#endregion
		}

		public interface CartesianSpatialOptionsFactory
		{
			#region spatial_search_enhancements_5
			SpatialOptions BoundingBoxIndex();

			SpatialOptions QuadPrefixTreeIndex(int maxTreeLevel, SpatialBounds bounds);

			#endregion
		}

		public interface Foo
		{
			#region spatial_search_0
			object SpatialGenerate(double lat, double lng);

			object SpatialGenerate(string fieldName, double lat, double lng);

			object SpatialGenerate(string fieldName, string shapeWKT);

			object SpatialGenerate(string fieldName, string shapeWKT, SpatialSearchStrategy strategy);

			object SpatialGenerate(string fieldName, string shapeWKT, SpatialSearchStrategy strategy, int maxTreeLevel);

			#endregion

			#region spatial_search_5
			IDocumentQueryCustomization RelatesToShape(string fieldName, string shapeWKT, SpatialRelation rel);

			#endregion
		}

		#region spatial_search_6
		public enum SpatialSearchStrategy
		{
			GeohashPrefixTree,
			QuadPrefixTree,
		}

		#endregion

		#region spatial_search_7
		public enum SpatialRelation
		{
			Within,
			Contains,
			Disjoint,
			Intersects,

			/// <summary>
			/// Does not filter the query, merely sort by the distance
			/// </summary>
			Nearby
		}

		#endregion
	}

	public class SpatialDoc
	{
		public object Shape { get; set; }

		public object Point { get; set; }
	}

	#region spatial_search_enhancements_8
	public class SpatialDoc_Index : AbstractIndexCreationTask<SpatialDoc>
	{
		public SpatialDoc_Index()
		{
			Map = docs => from spatial in docs
						  select new
						  {
							  Shape = spatial.Shape,
							  Point = spatial.Point
						  };

			Spatial(x => x.Shape, options => options.Geography.Default());
			Spatial(x => x.Point, options => options.Cartesian.BoundingBoxIndex());
		}
	}

	#endregion

	#region spatial_search_enhancements_1
	public class EventWithWKT
	{
		public string Id { get; set; }

		public string Name { get; set; }

		public string WKT { get; set; }
	}

	#endregion

	#region spatial_search_1
	public class Event
	{
		public string Id { get; set; }

		public string Name { get; set; }

		public double Latitude { get; set; }

		public double Longitude { get; set; }
	}

	#endregion

	#region spatial_search_2
	public class Events_SpatialIndex : AbstractIndexCreationTask<Event>
	{
		public Events_SpatialIndex()
		{
			Map = events => from e in events
							select new
							{
								Name = e.Name,
								__ = SpatialGenerate("Coordinates", e.Latitude, e.Longitude)
							};
		}
	}

	#endregion

	#region spatial_search_enhancements_2
	public class EventsWithWKT_SpatialIndex : AbstractIndexCreationTask<EventWithWKT>
	{
		public EventsWithWKT_SpatialIndex()
		{
			Map = events => from e in events
							select new
							{
								Name = e.Name,
								WKT = e.WKT
							};

			Spatial(x => x.WKT, options => options.Geography.Default());
		}
	}

	#endregion

	public class SpatialIndexes
	{
		public void Sample()
		{
			using (var store = new DocumentStore())
			{
				using (var session = store.OpenSession())
				{
					#region spatial_search_3
					session.Query<Event, Events_SpatialIndex>()
						   .Customize(x => x.WithinRadiusOf(
							   fieldName: "Coordinates",
							   radius: 10,
							   latitude: 32.1234,
							   longitude: 23.4321))
						   .ToList();
					#endregion

					#region spatial_search_8
					session.Advanced.LuceneQuery<Event, Events_SpatialIndex>()
						   .WithinRadiusOf(fieldName: "Coordinates", radius: 10, latitude: 32.1234, longitude: 23.4321)
						   .ToList();

					#endregion
				}

				using (var session = store.OpenSession())
				{
					#region spatial_search_4
					session.Query<Event, Events_SpatialIndex>()
							.Customize(x => x.RelatesToShape("Coordinates", "Circle(32.1234, 23.4321, d=10.0000)", SpatialRelation.Within))
							.ToList();

					#endregion

					#region spatial_search_9
					session.Advanced.LuceneQuery<Event, Events_SpatialIndex>()
						   .RelatesToShape("Coordinates", "Circle(32.1234, 23.4321, d=10.0000)", SpatialRelation.Within)
						   .ToList();

					#endregion

					#region spatial_search_enhancements_6
					var point = new
					{
						type = "Point",
						coordinates = new[] { -10d, 45d }
					};

					session.Store(new SpatialDoc { Shape = point });

					#endregion

					#region spatial_search_enhancements_7
					session.Store(new SpatialDoc { Point = new[] { -10d, 45d } });
					session.Store(new SpatialDoc { Point = new { X = -10d, Y = 45d } });
					session.Store(new SpatialDoc { Point = new { Latitude = 45d, Longitude = -10d } });
					session.Store(new SpatialDoc { Point = new { lat = 45d, lon = -10d } });
					session.Store(new SpatialDoc { Point = new { lat = 45d, lng = -10d } });
					session.Store(new SpatialDoc { Point = new { Lat = 45d, Long = -10d } });
					session.Store(new SpatialDoc { Point = "geo:45.0,-10.0;u=2.0" }); // Geo URI

					#endregion

					object someWktShape = null;

					#region spatial_search_enhancements_9
					session.Query<SpatialDoc, SpatialDoc_Index>()
						.Spatial(x => x.Shape, criteria => criteria.WithinRadiusOf(500, 30, 30))
						.ToList();

					session.Query<SpatialDoc, SpatialDoc_Index>()
						.Spatial(x => x.Shape, criteria => criteria.Intersects(someWktShape))
						.ToList();

					#endregion
				}
			}
		}
	}
}

File: Documentation/3.0/Samples/java/src/test/java/net/ravendb/indexes/SpatialIndexes.java
package net.ravendb.indexes;

import java.util.List;

import com.mysema.query.annotations.QueryEntity;

import net.ravendb.abstractions.basic.UseSharpEnum;
import net.ravendb.abstractions.indexing.SpatialOptions;
import net.ravendb.abstractions.indexing.SpatialOptions.SpatialRelation;
import net.ravendb.abstractions.indexing.SpatialOptions.SpatialUnits;
import net.ravendb.abstractions.indexing.SpatialOptionsFactory.CartesianSpatialOptionsFactory;
import net.ravendb.abstractions.indexing.SpatialOptionsFactory.GeographySpatialOptionsFactory;
import net.ravendb.abstractions.indexing.SpatialOptionsFactory.SpatialBounds;
import net.ravendb.client.IDocumentSession;
import net.ravendb.client.IDocumentStore;
import net.ravendb.client.document.DocumentQueryCustomizationFactory;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.indexes.AbstractIndexCreationTask;
import net.ravendb.client.spatial.SpatialCriteria;
import net.ravendb.client.spatial.SpatialCriteriaFactory;


public class SpatialIndexes {

  //region spatial_search_enhancements_3
  public static class SpatialOptionsFactory {
    public GeographySpatialOptionsFactory getGeography() {
      return new GeographySpatialOptionsFactory();
    }

    public CartesianSpatialOptionsFactory getCartesian() {
      return new CartesianSpatialOptionsFactory();
    }
  }
  //endregion

  public interface IGeographySpatialOptionsFactory {
    //region spatial_search_enhancements_4
    public SpatialOptions defaultOptions();

    // GeohashPrefixTree strategy with maxTreeLevel set to 9
    public SpatialOptions defaultOptions(SpatialUnits circleRadiusUnits);

    public SpatialOptions boundingBoxIndex();

    public SpatialOptions boundingBoxIndex(SpatialUnits circleRadiusUnits);

    public SpatialOptions geohashPrefixTreeIndex(int maxTreeLevel);

    public SpatialOptions geohashPrefixTreeIndex(int maxTreeLevel, SpatialUnits circleRadiusUnits);

    public SpatialOptions quadPrefixTreeIndex(int maxTreeLevel);

    public SpatialOptions quadPrefixTreeIndex(int maxTreeLevel, SpatialUnits circleRadiusUnits);
    //endregion
  }

  public interface ISpatialCriteriaFactory {
    //region spatial_search_enhancements_a
    public SpatialCriteria relatesToShape(Object shape, SpatialRelation relation);

    public SpatialCriteria intersects(Object shape);

    public SpatialCriteria contains(Object shape);

    public SpatialCriteria disjoint(Object shape);

    public SpatialCriteria within(Object shape);

    public SpatialCriteria withinRadiusOf(double radius, double x, double y);
    //endregion
  }

  public interface ICartesianSpatialOptionsFactory {
    //region spatial_search_enhancements_5
    public SpatialOptions boundingBoxIndex();

    public SpatialOptions quadPrefixTreeIndex(int maxTreeLevel, SpatialBounds bounds);
    //endregion
  }

  public interface Foo {
    //region spatial_search_5
    public DocumentQueryCustomizationFactory relatesToShape(final String fieldName, final String shapeWKT, final SpatialRelation rel);
    //endregion
  }

  //region spatial_search_7
  @UseSharpEnum
  public enum SpatialRelation {
    WITHIN, CONTAINS, DISJOINT, INTERSECTS,

    /**
     * Does not filter the query, merely sort by the distance
     */
    NEARBY
  }
  //endregion

  @QueryEntity
  public static class SpatialDoc {
    private Object shape;
    private Object point;

    public Object getShape() {
      return shape;
    }
    public void setShape(Object shape) {
      this.shape = shape;
    }
    public Object getPoint() {
      return point;
    }
    public void setPoint(Object point) {
      this.point = point;
    }
  }

  //region spatial_search_enhancements_8
  public static class SpatialDoc_ByShapeAndPoint extends AbstractIndexCreationTask {
    public SpatialDoc_ByShapeAndPoint() {
      QSpatialIndexes_SpatialDoc s = QSpatialIndexes_SpatialDoc.spatialDoc;
      map =
       " from spatial in docs        " +
       " select new                  " +
       " {                           " +
       "     Shape = spatial.Shape,  " +
       "     Point = spatial.Point   " +
       " }; ";

      spatial(s.shape, new SpatialOptionsFactory().getGeography().defaultOptions());
      spatial(s.point, new SpatialOptionsFactory().getCartesian().boundingBoxIndex());
    }
  }
  //endregion

  //region spatial_search_enhancements_1
  @QueryEntity
  public static class EventWithWKT {
    private String id;
    private String name;
    private String WKT;

    public String getId() {
      return id;
    }

    public void setId(String id) {
      this.id = id;
    }

    public String getName() {
      return name;
    }

    public void setName(String name) {
      this.name = name;
    }

    public String getWKT() {
      return WKT;
    }

    public void setWKT(String wKT) {
      WKT = wKT;
    }
  }
  //endregion

  //region spatial_search_1
  public static class Event {
    private String id;
    private String name;
    private double latiude;
    private double longitude;

    public String getId() {
      return id;
    }
    public void setId(String id) {
      this.id = id;
    }
    public String getName() {
      return name;
    }
    public void setName(String name) {
      this.name = name;
    }
    public double getLatiude() {
      return latiude;
    }
    public void setLatiude(double latiude) {
      this.latiude = latiude;
    }
    public double getLongitude() {
      return longitude;
    }
    public void setLongitude(double longitude) {
      this.longitude = longitude;
    }
  }
  //endregion

  //region spatial_search_2
  public static class Events_ByNameAndCoordinates extends AbstractIndexCreationTask {
    public Events_ByNameAndCoordinates() {
      map =
       " from e in docs.Events                                             " +
       " select new                                                        " +
       " {                                                                 " +
       "     Name = e.Name,                                                " +
       "     __ = SpatialGenerate(\"Coordinates\", e.Latitude, e.Longitude)" +
       " }; ";
    }
  }
  //endregion

  //region spatial_search_enhancements_2
  public static class EventsWithWKT_ByNameAndWKT extends AbstractIndexCreationTask {
    public EventsWithWKT_ByNameAndWKT() {
      QSpatialIndexes_EventWithWKT e = QSpatialIndexes_EventWithWKT.eventWithWKT;
      map =
       " from e in docs.Events    " +
       " select new          " +
       " {                   " +
       "     Name = e.Name,  " +
       "     WKT = e.WKT     " +
       " }; ";

      spatial(e.WKT, new SpatialOptionsFactory().getGeography().defaultOptions());
    }
  }
  //endregion

  @SuppressWarnings("unused")
  public void sample() throws Exception {


    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        //region spatial_search_3
        List<Event> results = session
          .query(Event.class, Events_ByNameAndCoordinates.class)
          .customize(new DocumentQueryCustomizationFactory().withinRadiusOf("Coordinates", 10, 32.1234, 23.4321))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region spatial_search_8
        List<Event> results = session
          .advanced()
          .documentQuery(Event.class, Events_ByNameAndCoordinates.class)
          .withinRadiusOf("Coordinates", 10, 32.1234, 23.4321)
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region spatial_search_4
        List<Event> results = session
          .query(Event.class, Events_ByNameAndCoordinates.class)
          .customize(new DocumentQueryCustomizationFactory()
            .relatesToShape("Coordinates", "Circle(32.1234 23.4321 d=10.0000)",
              net.ravendb.abstractions.indexing.SpatialOptions.SpatialRelation.WITHIN))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        //region spatial_search_9
        List<Event> results = session
          .advanced()
          .documentQuery(Event.class, Events_ByNameAndCoordinates.class)
          .relatesToShape("Coordinates", "Circle(32.1234 23.4321 d=10.0000)",
              net.ravendb.abstractions.indexing.SpatialOptions.SpatialRelation.WITHIN)
          .toList();
        //endregion
      }
    }

    try (IDocumentStore store = new DocumentStore()) {
      try (IDocumentSession session = store.openSession()) {
        Object someWktShape = null;

        //region spatial_search_enhancements_9
        QSpatialIndexes_SpatialDoc s = QSpatialIndexes_SpatialDoc.spatialDoc;
        List<SpatialDoc> results = session
          .query(SpatialDoc.class, SpatialDoc_ByShapeAndPoint.class)
          .spatial(s.shape, new SpatialCriteriaFactory().withinRadiusOf(500, 30, 30))
          .toList();
        //endregion
      }

      try (IDocumentSession session = store.openSession()) {
        Object someWktShape = null;

        //region spatial_search_enhancements_1_0
        QSpatialIndexes_SpatialDoc s = QSpatialIndexes_SpatialDoc.spatialDoc;
        List<SpatialDoc> results = session
          .advanced()
          .documentQuery(SpatialDoc.class, SpatialDoc_ByShapeAndPoint.class)
          .spatial(s.shape, new SpatialCriteriaFactory().intersects(someWktShape))
          .toList();
        //endregion
      }
    }
  }

}
